utils/measure/measure/util/measure_util.py

```diff
@@ -21,21 +26,113 @@ def __init__(self, power_meter: PowerMeter, config: MeasureConfig) -> None:
         self.dummy_load_value: float | None = None
         self.config = config
 
-    def take_average_measurement(self, duration: int) -> float:
-        """Measure average power consumption for a given time period in seconds"""
-        _LOGGER.info("Measuring average power for %d seconds", duration)
+    def take_average_measurement(self, duration: int, measure_resistance: bool = False) -> float:
```

Comment:
I have asked chatGPT to do a first refactor round to make things more readable.
It's improved a lot imo this way.
So I suggest to just use this mainly ;-)

```python
def take_average_measurement(self, duration: int, measure_resistance: bool = False) -> float:
    """Measures average power or resistance over a specified duration."""
    _LOGGER.info(
        "Measuring average %s over %s seconds",
        "resistance" if measure_resistance else "power",
        duration,
    )
    start_time = time.time()
    readings: list[float] = []

    while (time.time() - start_time) < duration:
        if not self._can_continue_measurement(start_time, duration):
            break

        time.sleep(self.config.sleep_time)
        measurement = self._perform_measurement(measure_resistance)
        
        if measurement is not None:
            readings.append(measurement)

    if not readings:
        _LOGGER.error("No valid readings were recorded.")
        exit(1)

    average = round(mean(readings), 2)
    unit = "Ω" if measure_resistance else "W"
    _LOGGER.info("Average of %d measurements: %.2f %s", len(readings), average, unit)
    return average

def _can_continue_measurement(self, start_time: float, duration: int) -> bool:
    """Checks if the measurement can continue based on elapsed time."""
    return (time.time() - start_time + self.config.sleep_time) < duration

def _perform_measurement(self, measure_resistance: bool) -> float | None:
    """Performs a single measurement and returns the result."""
    if measure_resistance:
        return self._measure_resistance()

    if self.dummy_load_value:
        return self._measure_power_with_dummy_load()

    return self._measure_power_without_dummy_load()

def _measure_resistance(self) -> float | None:
    """Measures resistance based on voltage and power."""
    result = self.power_meter.get_power(include_voltage=True)
    power, voltage = result.power, result.voltage

    if voltage < 1:
        _LOGGER.error("Error: Voltage measurement returned zero. Aborting measurement.")
        exit(1)

    if round(power, 2) == 0:
        _LOGGER.warning("Invalid measurement: power: %.2f W, voltage: %.2f", power, voltage)
        return None

    resistance = round((voltage**2) / power, 4)
    _LOGGER.debug("Measured resistance: %.2f Ω; power: %.2f W, voltage: %.2f", resistance, power, voltage)
    return resistance

def _measure_power_with_dummy_load(self) -> float | None:
    """Measures power with the dummy load value accounted for."""
    self._validate_voltage_support()
    result = self.power_meter.get_power(include_voltage=True)
    power, voltage = result.power, result.voltage

    if voltage < 1:
        _LOGGER.error("Error: Voltage measurement returned zero. Aborting measurement.")
        exit(1)

    dummy_power = (voltage**2) / self.dummy_load_value
    power -= dummy_power

    if round(power, 2) <= 0:
        _LOGGER.warning(
            "Invalid measurement after subtracting dummy load consumption. Calculated consumption: %.2f W; ignoring",
            power,
        )
        return None

    _LOGGER.info("Measured power: %.2f W", power)
    return power

def _measure_power_without_dummy_load(self) -> float | None:
    """Measures power without the dummy load."""
    power = self.power_meter.get_power().power

    if round(power, 2) == 0:
        _LOGGER.warning("Invalid measurement. Consumption: %.2f W; ignoring", power)
        return None

    _LOGGER.info("Measured power: %.2f W", power)
    return power
```

Reply:
Issue with ChatGPT is, that it will often introduce 10 oversights while doing this, so you have to careful read every change it does ... which defeats the purpose a bit, as this takes longer than actually writing the code ;)

Reply:
That's why I started with unit tests for the measure tool as well. It's undoable to constantly test manually.
Ideally we need to add tests for this new feature as well. I can also do that later when you don't have the time / don't want to take on the task.
Powercalc integration is covered 100% with tests, so I can safely do refactors without the fear to break stuff. In the end we need more tests for the measure tool as well.
 
95% of the time the refactor chatGPT does will be right and behave the same functionality wise.
I did a quick look through the code before pasting here and seems fine by me.

The problem with the current code is there are many code paths / high complexity.
Whenever there are bugs I need to investigate and it will take more time this way.
So we really must improve when introducing new features and not makes things worse.
Constantly refactoring is essential part of software development.

Reply:
Yeah, it's too complex, but cutting down complexity is out of scope of this PR.

I'm however happy to refactor in a different PR build on top of this the complexity down, if that's a compromise?